File: backend/app/models/schemas/audit.py
"""
Audit Log Pydantic schemas for API request/response validation.

Per TOR 7.1.6 requirement for access logs and audit trail.
"""

from datetime import datetime

from pydantic import BaseModel, Field


class AuditLogEntry(BaseModel):
    """Schema for audit log entry response."""

    id: int
    time: datetime
    user_id: str | None = None
    user_email: str | None = None
    user_ip: str | None = None
    action: str
    resource_type: str | None = None
    resource_id: str | None = None
    request_method: str | None = None
    request_path: str | None = None
    request_body: dict | None = None
    response_status: int | None = None
    user_agent: str | None = None
    session_id: str | None = None

    model_config = {"from_attributes": True}


class AuditLogFilter(BaseModel):
    """Schema for audit log filtering parameters."""

    user_id: str | None = Field(default=None, description="Filter by user ID")
    user_email: str | None = Field(default=None, description="Filter by user email")
    action: str | None = Field(
        default=None, description="Filter by action: read, create, update, delete"
    )
    resource_type: str | None = Field(
        default=None,
        description="Filter by resource type: forecast, alerts, data, etc.",
    )
    resource_id: str | None = Field(default=None, description="Filter by resource ID")
    request_method: str | None = Field(
        default=None, description="Filter by HTTP method: GET, POST, PUT, DELETE"
    )
    response_status: int | None = Field(
        default=None, description="Filter by HTTP status code"
    )
    start_date: datetime | None = Field(
        default=None, description="Start date for filtering"
    )
    end_date: datetime | None = Field(
        default=None, description="End date for filtering"
    )
    user_ip: str | None = Field(default=None, description="Filter by IP address")


class AuditLogStats(BaseModel):
    """Schema for audit log summary statistics."""

    total_requests: int
    unique_users: int
    successful_requests: int
    failed_requests: int
    actions_breakdown: dict[str, int]
    resources_breakdown: dict[str, int]
    top_users: list[dict[str, int]]
    error_rate: float
    period_start: datetime
    period_end: datetime


class AuditLogExport(BaseModel):
    """Schema for audit log export request."""

    format: str = Field(default="csv", description="Export format: csv, json")
    filters: AuditLogFilter | None = Field(
        default=None, description="Optional filters to apply"
    )
    include_request_body: bool = Field(
        default=False, description="Include request body in export"
    )


class AuditLogResponse(BaseModel):
    """Generic response wrapper for audit endpoints."""

    status: str
    data: dict


class UserActivitySummary(BaseModel):
    """Schema for user activity summary."""

    user_id: str
    user_email: str | None = None
    total_requests: int
    last_activity: datetime
    actions_breakdown: dict[str, int]
    most_accessed_resources: list[dict[str, int]]
    failed_requests: int
    avg_requests_per_day: float

File: frontend/src/types/audit.ts
/**
 * Audit Log Types for PEA RE Forecast Platform
 * TOR 7.1.6 Requirement: Security and Audit Trail
 */

export interface AuditLogEntry {
  id: number;
  time: string;
  user_id: string | null;
  user_email: string | null;
  user_ip: string | null;
  action: string;
  resource_type: string | null;
  resource_id: string | null;
  request_method: string | null;
  request_path: string | null;
  request_body: Record<string, unknown> | null;
  response_status: number | null;
  user_agent: string | null;
  session_id: string | null;
}

export interface AuditLogFilter {
  start_date?: string;
  end_date?: string;
  user_id?: string;
  action?: string;
  resource_type?: string;
  request_method?: string;
  response_status?: number;
}

export interface AuditLogStats {
  // API returns these fields
  total_requests: number;
  unique_users: number;
  successful_requests: number;
  failed_requests: number;
  actions_breakdown: Record<string, number>;
  resources_breakdown: Record<string, number>;
  top_users: Array<{
    user_id: string;
    count: number;
  }>;
  error_rate: number;
  period_start: string;
  period_end: string;
}

export interface AuditLogResponse {
  status: string;
  data: {
    logs: AuditLogEntry[];
    total: number;
    page: number;
    limit: number;
    pages: number;
  };
}

export interface AuditLogStatsResponse {
  status: string;
  data: {
    stats: AuditLogStats;
  };
}

export interface AuditLogExportRequest {
  filters: AuditLogFilter;
  format: "csv" | "json";
}

export type AuditAction =
  | "read"
  | "create"
  | "update"
  | "delete"
  | "login"
  | "logout"
  | "export"
  | "import"
  | "configure";

export type AuditResourceType =
  | "forecast"
  | "voltage"
  | "alert"
  | "model"
  | "user"
  | "settings"
  | "data"
  | "report";
